lectures/sympy.md

```diff
@@ -0,0 +1,992 @@
+
+```{code-cell} ipython3
+factor(expand_expr)
+```
+
+We can solve this expression
+
+```{code-cell} ipython3
+solve(expr)
+```
+
+Note this is equivalent to solving the following equation for `x`
+
+$$
+(x + y)^2 = 0 
+$$
+
+Another way to solve this equation is to use `solveset`.
+
+`solveset` allows users to define the domain of `x` in the evaluation and output a set as a solution
+
+```{code-cell} ipython3
+solveset(expr, x, domain=S.Naturals)
+```
+
+```{note}
+[Solvers](https://docs.sympy.org/latest/modules/solvers/index.html) is an important module with tools to solve different types of equations. 
+
+There are a variety of solvers available in SymPy depending on the nature of the problem.
+```
+
+### Equations
+
+SymPy provides several functions to manipulate equations.
+
+Let's develop an equation with the expression we defined before
+
+```{code-cell} ipython3
+eq = Eq(expr, 0)
+eq
+```
+
+Solving this equation with respect to $x$ gives the same output as solving the expression directly
+
+```{code-cell} ipython3
+solve(eq, x)
+```
+
+SymPy can handle equations with multiple solutions
+
+```{code-cell} ipython3
+eq = Eq(expr, 1)
+solve(eq, x)
+```
+
+`solve` function can also combine multiple equations together and solve a system of equations
+
+```{code-cell} ipython3
+eq2 = Eq(x, y)
+eq2
+```
+
+```{code-cell} ipython3
+solve([eq, eq2], [x, y])
+```
+
+We can also solve for the value of $y$ by simply substituting $x$ with $y$
+
+```{code-cell} ipython3
+expr_sub = expr.subs(x, y)
+expr_sub
+```
+
+```{code-cell} ipython3
+solve(Eq(expr_sub, 1))
+```
+
+Below is another example equation with the symbol `x` and functions `sin`, `cos`, and `tan` using the `Eq` function
+
+```{code-cell} ipython3
+# Create an equation
+eq = Eq(cos(x) / (tan(x)/sin(x)), 0)
+eq
+```
+
+Now we simplify this equation using the `simplify` function
+
+```{code-cell} ipython3
+# Simplify an expression
+simplified_expr = simplify(eq)
+simplified_expr
+```
+
+Again, we use the `solve` function to solve this equation
+
+```{code-cell} ipython3
+# Solve the equation
+sol = solve(eq, x)
+sol
+```
+
+SymPy can also handle more complex equations involving trigonometry and complex numbers.
+
+We demonstrate this using [Euler's formula](https://en.wikipedia.org/wiki/Euler%27s_formula)
+
+```{code-cell} ipython3
+# 'I' represents the imaginary number i 
+euler = cos(x) + I*sin(x)
+euler
+```
+
+```{code-cell} ipython3
+simplify(euler)
+```
+
+If you are interested, we encourage you to read the lecture on [trigonometry and complex numbers](https://python.quantecon.org/complex_and_trig.html).
+
+#### Example: fixed point computation
+
+Fixed point computation is frequently used in economics and finance.
+
+Here we solve the fixed point of the Solow-Swan growth dynamics:
+
+$$
+k_{t+1}=s f\left(k_t\right)+(1-\delta) k_t, \quad t=0,1, \ldots
+$$
+
+where $k_t$ is the capital stock, $f$ is a production function, $\delta$ is a rate of depreciation.
+
```

Comment:
Perhaps it's a good idea to write down the equation being solved (for people who aren't completely sure what a fixed point is.